src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsProjectJsonMigrator.cs

```diff
@@ -0,0 +1,51 @@
+﻿using NuGet.PackageManagement.UI;
+using NuGet.PackageManagement.VisualStudio;
+using System;
+using System.ComponentModel.Composition;
+using System.Threading.Tasks;
+
+namespace NuGet.VisualStudio
+{
+    [Export(typeof(IVsProjectJsonMigrator))]
+    class VsProjectJsonMigrator : IVsProjectJsonMigrator
+    {
+        private IVsSolutionManager _solutionManager;
+        private PumpingJTF _pumpingJtf;
+
+        [ImportingConstructor]
+        public VsProjectJsonMigrator()
+        {
+            _solutionManager = ServiceLocator.GetInstance<IVsSolutionManager>();
+            _pumpingJtf = new PumpingJTF(NuGetUIThreadHelper.JoinableTaskFactory.Context);
+        }
+        public IVsProjectJsonMigrateResult MigrateProjectToPackageRef(string projectUniqueName)
+        {            
+            return _pumpingJtf.Run(async delegate
```

Comment:
Inside the JTF this should call TaskScheduler.Default to move to the background then, Then calls that use DTE should again use a JTF and call Move to UI thread so that only necessary actions happen on the UI thread. Without that you will see VS lock up until this finishes.